runtime/protos/runtime.proto

```diff
@@ -17,232 +17,270 @@ option go_package = "github.com/ServiceWeaver/weaver/runtime/protos";
 
-  // Latest log entry.
-  LogEntry log_entry = 3;
+  // Envelope initiated RPC replies.
+  GetHealthReply get_health_reply = 100;
+  GetMetricsReply get_metrics_reply = 101;
+  GetLoadReply get_load_reply = 102;
+  GetProfileReply get_profile_reply = 103;
+  UpdateRoutingInfoReply update_routing_info_reply = 104;
+  UpdateComponentsReply update_components_reply = 105;
 
-  // Trace spans to be exported.
-  Spans trace_spans = 4;
+  // Weavelet initiated RPC requests.
+  EnsureComponentRequest ensure_component_request = 201;
+  GetAddressRequest get_address_request = 201;
+  ExportListenerRequest export_listener_request = 202;
 
-  // Component to be started.
-  ComponentToStart component_to_start = 5;
+  // Weavelet initiated messages.
+  LogEntry log_entry = 301;
+  TraceSpans trace_spans = 302;
+}
 
-  // Weavelet load to be reported.
-  WeaveletLoadReport load_report = 8;
+// EnvelopeInfo is the information provided by an envelope to a weavelet during
+// the initial envelope-weavelet handshake.
+message EnvelopeInfo {
+  string app = 1;                    // app name
+  string deployment_id = 2;          // deployment id, in UUID format
+  string id = 3;                     // weavelet id, in UUID format
+  map<string, string> sections = 4;  // See AppConfig.Sections.
+  bool single_process = 5;           // a single-process deployment?
+  bool single_machine = 6;           // a single-machine deployment?
+  bool run_main = 7;                 // run the main function?
+}
 
-  // Updated health status.
-  HealthReport health_report = 9;
+// WeaveletInfo is the information provided by a weavelet to an envelope during
+// the initial envelope-weavelet handshake.
+message WeaveletInfo {
+  // Internal TCP address of the weavelet. This address will be passed to other
+  // weavelets as part of their RoutingInfo.
+  string dial_addr = 1;
 
-  // Weavelet profile.
-  Profile profile = 10;
+  // Pid for the weavelet OS process.
+  //
+  // TODO(mwhittaker): Remove pid.
+  int64 pid = 2;
+}
 
-  // Information sent back by the weavelet upon startup.
-  WeaveletInfo weavelet_info = 16;
+// GetHealthRequest is a request from an envelope for a weavelet's health.
+//
+// Note that an envelope runs a weavelet in a subprocess, and the envelope can
+// tell whether the subprocess is alive without having to communicate over the
+// pipe. However, the envelope must communicate with the weavelet to check if it
+// is responsive.
+message GetHealthRequest {}
+
+// GetHealthReply is a reply to a GetHealthRequest.
+message GetHealthReply {
+  HealthStatus status = 1;
+}
 
-  // A reply to the envelope with an error status.
-  string error = 11;
+// HealthStatus specifies the health of a weavelet.
+enum HealthStatus {
+  UNKNOWN = 0;
+  HEALTHY = 1;
+  UNHEALTHY = 2;
+  TERMINATED = 3;
+}
 
-  // Request to get the address for a listener.
-  GetAddressRequest get_address_request = 14;
+// GetMetricsRequest is a request from an envelope for a weavelet's metrics.
```

Comment:
Can there be only one outstanding GetMetricsRequest at a time? That would maybe make sense, given the "changed metrics" semantics below?